### ISSUE TITLE
Add procedural macro for simpler yield syntax

### ISSUE THREAD

**Issue by runfalk** [whatisaphone/genawaiter#2]
Thank you for writing this!

The current yield syntax is a bit clunky and in my opinion a procedural macro could improve this.

It could look something like this:

```rust
#[sync::generator]
fn my_gen() -> impl Generator<Yield = u32, Return = bool> {
    yield 1;
    yield 2;
    yield 3;
    true
}
```

This would be translated into something like:

```rust
fn my_gen() -> genawaiter::sync::Gen<u32, bool, Future<...>> {
    Gen::new(|co| async move {
        co.yield_(1).await;
        co.yield_(2).await;
        co.yield_(3).await;
        true
    })
}

**Comment by HadrienG2** [comment]
I previously proposed something similar in [the urlo announcement](https://users.rust-lang.org/t/announcing-genawaiter-use-generators-yield-on-stable-rust/34283), but @whatisaphone was not enthusiastic about it because he thought use of macros would make things needlessly magical.

Nothing prevents you from implementing such a macro as a separate crate though.

**Comment by whatisaphone** [comment]
Well, if enough people ask for this, eventually there comes a point where I can't say no :)

I haven't explored the design space at all and I think it's going to take some experimentation to find the right set of tradeoffs. For example, all the proposals I've seen so far have used `fn` syntax. What about closures and capturing variables?

Fair warning, I'm planning on working on #1 before this, but I'll try to get to this at some point too.

**Comment by runfalk** [comment]
What about a function-like procedural macro for closures?

```rust
let gen = sync::generator!(|arg1, arg2| {
    yield arg1;
    yield arg2;
});
```

This would desugar into:

```rust
let gen = |arg1, arg2| {
    Gen::new(|co| async move {
        co.yield_(arg1).await;
        co.yield_(arg2).await;
    })
};
```

**Comment by vi** [comment]
Shall `yield_` be `r#yield` by the way? Which one is more idiomatic?

**Comment by HadrienG2** [comment]
I prefer yield_, it's one character shorter and raw identifier syntax may be unfamiliar to many users.

**Comment by runfalk** [comment]
I think `yield_` is more idiomatic. The most official source I can find is https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/advice.md.

> When a name is forbidden because it is a reserved word (e.g., crate), use a trailing underscore to make the name legal (e.g., crate_), or use raw identifiers if possible.

I see raw identifiers as something used for FFIs or automated tools to provide a migration path between Rust editions

**Comment by DevinR528** [comment]
I whipped up a `proc_macro_attr` for this it works on both `fn`'s and `Fn`'s (closures) I still have a few kinks to work out but will probably be done and documented in a day or so.

I can either publish it as a separate crate or start a PR to include it in here. [Here it is](https://github.com/DevinR528/yield-await)

An Example
```rust
use genawaiter::{
    generator_mut,
    stack::{self, Co, Gen, Shelf},
};

use gen_yielder::{yielder_cls, yielder_fn};

#[yielder_fn]
async fn odds() {
    for n in (1..).step_by(2).take_while(|&n| n < 10) {
        yield n
    }
}

fn main() {
    #[yielder_cls]
    let hello = async || {
        yield "hello"
    };
    generator_mut!(gen, hello);

    #[yielder_cls]
    let generator = Gen::new(|| async move {
        let mut n = 1;
        while n < 10 {
            // Suspend a function at any point with a value.
            yield n;
            n += 2;
        }
    });
    generator_mut!(gen, odds);

    for x in gen {
        println!("{}", x);
    }
}
```

**Comment by whatisaphone** [comment]
This looks great, thanks for putting it together! Would be awesome if you started cooking up a PR.

I haven't reviewed your code yet but right away I have a question – does the stable compiler really accept `yield n` syntax? I would have thought it would bail, even in a proc macro, but I'd be pleasantly surprised if that syntax works

**Comment by DevinR528** [comment]
Nope you're right, ughh I always forget I default my compiler to nightly. After actually implementing this I feel that maybe the readability and cleanliness gained from a `proc_macro` like this is far out weighed by the complexity and corner cases it creates. In order to use this for closures you need to also use`#![feature(proc_macro_hygiene)]` which I guess doesn't really matter if you already have to use `generators`. That being said I'd be happy to start a PR (I enjoy working on proc-macros) but this would seem to go against the very idea of the crate (stay off nightly).

**Comment by whatisaphone** [comment]
>the type signature of Co being different in the various mod's

Ah yes, I can see how that might be a problem. Your idea of three proc macros is a good one. genawaiter has some code shared between the different generators in its `core` module—that might be a good pattern to follow.

As for working on stable, I have a few ideas:

**Syntax:** In the original forum thread, HadrienG suggested `yield_!(x)` instead of `yield x`. That solves the syntax problem. (or you could be super cheeky and go with `x.yield_`)

**Closures:** A function-like macro could be used instead, and they are stable. Unfortunately the userland code ends up a little uglier. But hey, this crate is about stability, not looking pretty 😉

**Comment by whatisaphone** [comment]
This is implemented in v0.99
